File: 10 Days of Statistics/Day 5: Normal Distribution I.R
# https://www.hackerrank.com/challenges/s10-normal-distribution-1/problem

# erf(z) can be calculated using (2 * pnorm(x * sqrt(2)) - 1)

con <- file("stdin", open = "r")

dataset <- unlist(strsplit(readLines(con), split="\\s"))

# mean for X
X_mean <- as.numeric(dataset[1])
# standard deviation for X
X_sd  <- as.numeric(dataset[2])

# Q1, what is the probability that a car can be assembled at this plant in less than
Q1_time   <- as.numeric(dataset[3])

# Q2, what is the probability that a car can be assembled at this plant
#Between
Q2_time_lower <- as.numeric(dataset[4])
# and
Q2_time_upper <- as.numeric(dataset[5])

# Q1 probability
Q1_prob <- pnorm(Q1_time, mean = X_mean, sd = X_sd)

# Q2 probability lower 
Q2_prob_lower <- pnorm(Q2_time_lower, mean = X_mean, sd = X_sd)
# Q2 probability upper
Q2_prob_upper <- pnorm(Q2_time_upper, mean = X_mean, sd = X_sd)
# Q2 upper - lower to get the asked probability
Q2_prob <- Q2_prob_upper - Q2_prob_lower

cat(round(Q1_prob, digits=3))
cat("\n")
cat(round(Q2_prob, digits=3))

close(con)

File: 10 Days of Statistics/Day 5: Normal Distribution I.py
# https://www.hackerrank.com/challenges/s10-normal-distribution-1/problem

"""
Objective
In this challenge, we learn about normal distributions.

Task
In a certain plant, the time taken to assemble a car is a random variable, X, having a normal distribution with a mean of 20 hours and a standard deviation of 2 hours.
What is the probability that a car can be assembled at this plant in:
*Less than 19.5 hours?
*Between 20 and 22 hours?
"""

from math import sqrt, erf

# function to calculate the Cumulative probability
def cum_prob(X_mean, X_sd, x):
    erf_parameter = (x - X_mean) / (X_sd * sqrt(2))
    return 0.5 * (1 + erf(erf_parameter))

# respective mean and standard deviation for X
X_mean, X_sd = map(float, input().strip().split())

# Q1, what is the probability that a car can be assembled at this plant in less than
Q1_time = float(input().strip())

# Q2, what is the probability that a car can be assembled at this plant Between
Q2_time_lower, Q2_time_upper = map(float, input().strip().split())

# Q1 probability
Q1_prob = cum_prob(X_mean, X_sd, Q1_time)

# Q2 probability lower 
Q2_prob_lower = cum_prob(X_mean, X_sd, Q2_time_lower)
# Q2 probability upper
Q2_prob_upper = cum_prob(X_mean, X_sd, Q2_time_upper)
# Q2 upper - lower to get the asked probability
Q2_prob = Q2_prob_upper - Q2_prob_lower

print(round(Q1_prob,3))
print(round(Q2_prob,3))
